File: bolao/shared/scripts/notification_outbox.py
"""
notification_outbox.py — Python port of bolao/shared/scripts/notification_outbox.mjs, same JSON
schema and same idempotency-key format (football-hardening checkpoint F).

WHY THIS EXISTS: checkpoint D built the shared match-state-machine/outbox/reconciler pipeline in
JS. Checkpoint F's job is to actually WIRE that pipeline into each app's real operational flow —
but Copa2026's and CDB2026's real result-email automation (send_result_email.py) is Python, not
JS, and is a live, money-critical, audit-gated script (CLAUDE.md: "refuses to send any email if
[audit_scoring.py's self-test] fails"). Rewriting it wholesale to call into Node would be exactly
the kind of large, risky, non-reversible change CLAUDE.md's platform governance prohibits for a
script in production ("patches pequenos, testados e reversíveis").

So instead: this module speaks the EXACT SAME on-disk JSON contract as notification_outbox.mjs
(same field names: jobId, app, matchId, recipient, resultVersion, payloadSnapshot,
idempotencyKey, status, attemptCount, ...) so a job enqueued by one language is readable, claimable,
and completable by the other. That's the real integration point: both languages share ONE outbox,
enforcing the SAME duplicate-prevention guarantee, without either language needing to call into
the other's runtime.

send_result_email.py (both apps) gets one new, additive, off-by-default-until-called function:
check_idempotent_before_send() — a duplicate-prevention safety net that can only ever SKIP a send
that's a literal repeat of an already-recorded idempotency key; it can never suppress a send that
would have gone out before this change existed. See each script's own integration comment.
"""
from __future__ import annotations

import json
import os
import time
import random
import string
from datetime import datetime, timezone
from typing import Optional


def _now_iso() -> str:
    return datetime.now(timezone.utc).isoformat(timespec="milliseconds").replace("+00:00", "Z")


def default_outbox_path() -> str:
    return os.path.join(os.path.dirname(os.path.abspath(__file__)), "notification_outbox.json")


def read_all(path: str) -> list:
    if not os.path.exists(path):
        return []
    try:
        with open(path, "r", encoding="utf-8") as f:
            return json.load(f)
    except (json.JSONDecodeError, OSError):
        return []


def _write_all(jobs: list, path: str) -> None:
    tmp = f"{path}.tmp-{os.getpid()}-{''.join(random.choices(string.ascii_lowercase + string.digits, k=6))}"
    with open(tmp, "w", encoding="utf-8") as f:
        json.dump(jobs, f, ensure_ascii=False, indent=2)
        f.write("\n")
    os.replace(tmp, path)  # atomic on POSIX/Windows, same discipline as espn_provider.py


def idempotency_key(app: str, match_id: str, recipient: str, result_version) -> str:
    # MUST stay byte-for-byte identical to notification_outbox.mjs's idempotencyKey() format —
    # this is the entire cross-language interoperability contract. Covered by
    # test_notification_outbox_interop.mjs, which enqueues from Python and reads from Node (and
    # vice versa) and asserts the keys collide correctly.
    return f"{app}:{match_id}:{recipient}:v{result_version}"


def find_by_idempotency_key(key: str, path: Optional[str] = None) -> Optional[dict]:
    path = path or default_outbox_path()
    for job in read_all(path):
        if job.get("idempotencyKey") == key:
            return job
    return None


def enqueue(job: dict, path: Optional[str] = None) -> tuple[dict, bool]:
    """Same contract as notification_outbox.mjs's enqueue(): returns (job, created). A job with
    an idempotencyKey that already exists on disk (written by either language) is returned
    unchanged, created=False — this IS the duplicate-prevention mechanism."""
    path = path or default_outbox_path()
    jobs = read_all(path)
    existing = next((j for j in jobs if j.get("idempotencyKey") == job["idempotencyKey"]), None)
    if existing:
        return existing, False
    record = {
        "jobId": f"job_{int(time.time() * 1000)}_{''.join(random.choices(string.ascii_lowercase + string.digits, k=8))}",
        "app": job["app"],
        "matchId": job["matchId"],
        "recipient": job["recipient"],
        "resultVersion": job["resultVersion"],
        "payloadSnapshot": job["payloadSnapshot"],
        "idempotencyKey": job["idempotencyKey"],
        "status": "pending",
        "attemptCount": 0,
        "maxAttempts": job.get("maxAttempts", 5),
        "processingStartedAt": None,
        "lastAttemptAt": None,
        "sentAt": None,
        "lastError": None,
        "createdAt": _now_iso(),
    }
    jobs.append(record)
    _write_all(jobs, path)
    return record, True


def record_result(job_id: str, ok: bool, error: Optional[str] = None, path: Optional[str] = None) -> dict:
    path = path or default_outbox_path()
    jobs = read_all(path)
    idx = next((i for i, j in enumerate(jobs) if j.get("jobId") == job_id), None)
    if idx is None:
        raise ValueError(f"unknown jobId: {job_id}")
    j = jobs[idx]
    j["attemptCount"] = j.get("attemptCount", 0) + 1
    j["lastAttemptAt"] = _now_iso()
    j["processingStartedAt"] = None
    j["status"] = "sent" if ok else "failed"
    j["lastError"] = None if ok else (error or "unknown error")
    if ok:
        j["sentAt"] = j["lastAttemptAt"]
    jobs[idx] = j
    _write_all(jobs, path)
    return j

File: bolao/shared/scripts/notification_outbox.mjs
// notification_outbox.mjs — persisted (file-backed, NOT localStorage/in-memory-only) idempotent
// job store for match-result notification emails (football-hardening checkpoint D).
//
// Reuses the proven pattern from bolao/loterias/powerball/scripts/email/outbox.mjs (commit
// a3589c3, "feat(powerball): add email snapshots and idempotent outbox"): one JSON file per
// environment, one job per recipient, idempotency key lookup before insert, a frozen
// payloadSnapshot that retries replay verbatim rather than recompute. Extended here with:
//   - an explicit "processing" lock state (so two concurrent reconciler runs can't both pick up
//     the same pending job) with a processingStartedAt timestamp the reconciler uses to detect
//     and recover stuck jobs;
//   - a `resultVersion` folded into the idempotency key, so a correction to an already-notified
//     result creates NEW jobs (new version) rather than mutating/resending the old ones in place.
//
// Every timestamp comes from an injected `clock` (see fake_clock.mjs) — never Date.now() — so
// callers can run fully deterministic tests with no real sleep/wall-clock dependency.

import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";

const __dirname = path.dirname(fileURLToPath(import.meta.url));

export function defaultOutboxPath() {
  return path.join(__dirname, "notification_outbox.json");
}

export function readAll(file) {
  if (!fs.existsSync(file)) return [];
  try { return JSON.parse(fs.readFileSync(file, "utf8")); } catch { return []; }
}

function writeAll(jobs, file) {
  // Atomic-ish write: temp file + rename, same discipline as espn_provider.py's
  // write_snapshot_atomic() — a crash mid-write must never leave a corrupt/partial outbox file
  // that a concurrent reconciler run would read as truth.
  const tmp = `${file}.tmp-${process.pid}-${Math.random().toString(36).slice(2, 8)}`;
  fs.writeFileSync(tmp, JSON.stringify(jobs, null, 2) + "\n");
  fs.renameSync(tmp, file);
}

export function idempotencyKey(app, matchId, recipient, resultVersion) {
  return `${app}:${matchId}:${recipient}:v${resultVersion}`;
}

/** Enqueues one job per recipient, unless a job with the SAME idempotencyKey already exists —
 * that is the whole duplicate-prevention contract. Returns { job, created }. A duplicated
 * workflow run (same matchId/recipient/resultVersion) calling this twice returns the SAME job
 * the second time, created:false, and writes nothing new to disk. */
export function enqueue(job, clock, file = defaultOutboxPath()) {
  const jobs = readAll(file);
  const existing = jobs.find((j) => j.idempotencyKey === job.idempotencyKey);
  if (existing) return { job: existing, created: false };
  const record = {
    jobId: `job_${clock.nowMs()}_${Math.random().toString(36).slice(2, 10)}`,
    app: job.app,
    matchId: job.matchId,
    recipient: job.recipient,
    resultVersion: job.resultVersion,
    payloadSnapshot: job.payloadSnapshot, // frozen at enqueue time — retries NEVER recompute this
    idempotencyKey: job.idempotencyKey,
    status: "pending", // pending -> processing -> sent | failed (failed is retryable up to maxAttempts)
    attemptCount: 0,
    maxAttempts: job.maxAttempts ?? 5,
    processingStartedAt: null,
    lastAttemptAt: null,
    sentAt: null,
    lastError: null,
    createdAt: clock.nowIso(),
  };
  jobs.push(record);
  writeAll(jobs, file);
  return { job: record, created: true };
}

/** Atomically claims a pending (or previously-failed-and-retryable) job for processing. Returns
 * null if the job doesn't exist, is already terminal (sent), or is already being processed by
 * someone else — this is what makes two concurrent executions safe: only one of them can
 * successfully flip a given job from pending to processing. */
export function claimForProcessing(jobId, clock, file = defaultOutboxPath()) {
  const jobs = readAll(file);
  const idx = jobs.findIndex((j) => j.jobId === jobId);
  if (idx === -1) return null;
  const j = jobs[idx];
  if (j.status === "sent") return null; // already done — never re-send
  if (j.status === "processing") return null; // someone else already has the lock
  if (j.status === "failed" && j.attemptCount >= j.maxAttempts) return null; // exhausted
  j.status = "processing";
  j.processingStartedAt = clock.nowIso();
  jobs[idx] = j;
  writeAll(jobs, file);
  return j;
}

export function recordResult(jobId, result, clock, file = defaultOutboxPath()) {
  const jobs = readAll(file);
  const idx = jobs.findIndex((j) => j.jobId === jobId);
  if (idx === -1) throw new Error(`unknown jobId: ${jobId}`);
  const j = jobs[idx];
  j.attemptCount += 1;
  j.lastAttemptAt = clock.nowIso();
  j.processingStartedAt = null;
  j.status = result.ok ? "sent" : "failed";
  j.lastError = result.ok ? null : (result.error || "unknown error");
  if (result.ok) j.sentAt = j.lastAttemptAt;
  jobs[idx] = j;
  writeAll(jobs, file);
  return j;
}

/** Recovers jobs stuck in "processing" for longer than `thresholdMs` (e.g. the worker that
 * claimed them crashed / the process was killed mid-send) back to "pending" so the reconciler
 * can retry them. Never touches a job that's genuinely still within its processing window, and
 * never touches a job that already reached a terminal "sent" state. */
export function recoverStuckJobs(thresholdMs, clock, file = defaultOutboxPath()) {
  const jobs = readAll(file);
  const nowMs = clock.nowMs();
  let recovered = 0;
  for (const j of jobs) {
    if (j.status !== "processing" || !j.processingStartedAt) continue;
    const startedMs = new Date(j.processingStartedAt).getTime();
    if (nowMs - startedMs >= thresholdMs) {
      j.status = "pending";
      j.processingStartedAt = null;
      j.lastError = `recovered from stuck 'processing' state after ${nowMs - startedMs}ms`;
      recovered += 1;
    }
  }
  if (recovered > 0) writeAll(jobs, file);
  return recovered;
}

export function findByIdempotencyKey(key, file = defaultOutboxPath()) {
  return readAll(file).find((j) => j.idempotencyKey === key) || null;
}

export function jobsForMatch(matchId, file = defaultOutboxPath()) {
  return readAll(file).filter((j) => j.matchId === matchId);
}

export function retryableFailed(file = defaultOutboxPath()) {
  return readAll(file).filter((j) => j.status === "failed" && j.attemptCount < j.maxAttempts);
}
